crates/codegen/schema/src/types/parser.rs

```diff
@@ -20,9 +22,9 @@ pub enum ParserDefinition {
 
     #[schemars(title = "DelimitedBy Expression")]
     DelimitedBy {
-        open: String,
+        open: Reference,
```

Comment:
Using a single type for all also makes it easier for all users of this code, since there is only one type to convert, visit, validate, or match.

Reply:
Because they need to be scanners, but also with a name. Scanners have no internal names, so the name comes from the reference. They can`t be ParserRef, nor Scanner.

Reply:
Discussed face to face:

In general, I agree with the goal of having restrictive types, especially in the case of the new `Production` enum, where the benefits are across the board, and also benefiting validation/visitors/rewriters/pattern matches.

But in the case of `Reference, I'm trying to compare the pros/cons of each:

- If we use a separate `Reference` type, we get better intellisense, but at the expense of adding code to handle that new type everywhere. In visitors, pattern matchers, writers/serializers, and any other codegen output.
- If we use `ParserRef` instead, and post validate it is only `Parser::Reference`, then we get much simpler code everywhere, at the expense of noisy suggestions in intellisense.
- Both will have "roughly" the same validation code to check that the `reference:` points to a valid scanner or a parser.

I argue that intellisense of the schema is not that important of a goal right now, given that by the time we think about externalizing slang, it will probably have its own LSP anyways, or we might change the syntax.
I think it is not worth the cost of additional types, given that it can complicate the code, and introduce subtle bugs, like the ones we just found today: [[1]](https://github.com/NomicFoundation/slang/pull/332#discussion_r1130506990), [[2]](https://github.com/NomicFoundation/slang/pull/344#discussion_r1127710891), and [[3]](https://github.com/NomicFoundation/slang/pull/344#discussion_r1130509107).

Just my two cents, but the decision is up to you.
If you see otherwise, I'm fine with resolving as-is for now.

Reply:
I agree with the concern, but think that restriction via types is better than restriction via validation.

Reply:
And  i would thus like to introduce typed references e.g. `scanner:` and `parser:`.